prefixset_iter_test.go

```diff
@@ -0,0 +1,84 @@
+//go:build go1.23
+
+package netipds
+
+import (
+	"iter"
+	"net/netip"
+	"slices"
+	"testing"
+)
+
+func TestPrefixSetPrefixSeq(t *testing.T) {
+	tests := []struct {
+		add  []netip.Prefix
+		want []netip.Prefix
+	}{
+		{pfxs(), pfxs()},
+		{pfxs("::0/128"), pfxs("::0/128")},
+		{pfxs("::0/128", "::1/128"), pfxs("::0/128", "::1/128")},
+		{pfxs("::0/127", "::0/128"), pfxs("::0/127", "::0/128")},
+		{pfxs("::0/126", "::0/127"), pfxs("::0/126", "::0/127")},
+		{pfxs("::0/1", "::0/128"), pfxs("::0/1", "::0/128")},
+		{pfxs("0::0/127", "::0/128", "::1/128"), pfxs("0::0/127", "::0/128", "::1/128")},
+		{pfxs("0::0/127", "::0/128", "::2/128"), pfxs("0::0/127", "::0/128", "::2/128")},
+		{pfxs("1.2.3.0/24", "1.2.3.4/32"), pfxs("1.2.3.0/24", "1.2.3.4/32")},
+	}
+	for _, tt := range tests {
+		psb := &PrefixSetBuilder{}
+		for _, p := range tt.add {
+			psb.Add(p)
+		}
+		ps := psb.PrefixSet()
+		seq := ps.PrefixSeq()
+		checkPrefixSeq(t, ps.PrefixSeq(), tt.want)
+		checkYieldFalse(t, seq)
+	}
+}
+
+func TestPrefixSetPrefixCompactSeq(t *testing.T) {
+	tests := []struct {
+		add  []netip.Prefix
+		want []netip.Prefix
+	}{
+		{pfxs(), pfxs()},
+		{pfxs("::0/128"), pfxs("::0/128")},
+		{pfxs("::0/128", "::1/128"), pfxs("::0/128", "::1/128")},
+		{pfxs("::0/127", "::0/128"), pfxs("::0/127")},
+		{pfxs("::0/126", "::0/127"), pfxs("::0/126")},
+		{pfxs("::0/1", "::0/128"), pfxs("::0/1")},
+		{pfxs("0::0/127", "::0/128", "::1/128"), pfxs("::0/127")},
+		{pfxs("0::0/127", "::0/128", "::2/128"), pfxs("::0/127", "::2/128")},
+		{pfxs("1.2.3.0/24", "1.2.3.4/32"), pfxs("1.2.3.0/24")},
+	}
+	for _, tt := range tests {
+		psb := &PrefixSetBuilder{}
+		for _, p := range tt.add {
+			psb.Add(p)
+		}
+		ps := psb.PrefixSet()
+		seq := ps.PrefixCompactSeq()
+		checkPrefixSeq(t, seq, tt.want)
+		checkYieldFalse(t, seq)
+	}
+}
+
+func checkPrefixSeq(t *testing.T, seq iter.Seq[netip.Prefix], want []netip.Prefix) {
+	t.Helper()
+	got := slices.AppendSeq(make([]netip.Prefix, 0, len(want)), seq)
+	if !slices.Equal(got, want) {
+		t.Errorf("got %v, want %v", got, want)
+	}
+}
+
+func checkYieldFalse[T any](t *testing.T, seq iter.Seq[T]) {
+	t.Helper()
+	var i int
+	for range seq {
+		i++
+		break
+	}
+	if i > 1 {
```

Comment:
I don't have much experience with iterators, but shouldn't this be `i > 0`?

The `slices.AppendSeq` in `checkPrefixSeq` should exhaust the iterator, such that it no longer yields anything.

I think the implementation of the iterator needs to be adjusted - I'll comment there.

Reply:
> I don't have much experience with iterators, but shouldn't this be `i > 0`?

No. `checkYieldFalse` is meant to verify that the `yield` function is not called again after it returns `false`. In practice this is a runtime panic, and `checkYieldFalse` is meant to trigger the panic in case the iterator implementation is wrong.